src/main/java/io/jenkins/plugins/folderauth/FolderBasedAuthorizationStrategy.java

```diff
@@ -113,18 +127,26 @@ public GlobalAclImpl getRootACL() {
     }
 
     /**
-     * Used to initialize transient fields when loaded from disk
+     * Initializes the {@link FolderBasedAuthorizationStrategy} when read from the disk.
      *
-     * @return {@code this}
+     * @return a fully resolved {@link FolderBasedAuthorizationStrategy}
      */
     @Nonnull
     @SuppressWarnings("unused")
-    protected Object readResolve() {
-        jobAcls = new ConcurrentHashMap<>();
-        initCache();
-        generateNewGlobalAcl();
-        updateJobAcls(true);
-        return this;
+    private FolderBasedAuthorizationStrategy readResolve() {
+        return new FolderBasedAuthorizationStrategy(globalRoles, folderRoles);
+    }
+
+    /**
+     * Uses {@link #FolderBasedAuthorizationStrategy(HashSet, HashSet)} to
+     * create a proxy {@link FolderBasedAuthorizationStrategy} to be serialized.
+     *
+     * @return a proxy {@link FolderBasedAuthorizationStrategy} only for serialization.
+     */
+    @Nonnull
+    @SuppressWarnings("unused")
+    private FolderBasedAuthorizationStrategy writeReplace() {
```

Comment:
Not sure what purpose this is serving. You are returning an object with the same persisted fields as `this`, so far as I can tell. And your constructor is still using `ConcurrentHashMap.newKeySet()` for the persisted fields.

Reply:
Hi @jglick ! This creates a new object using a private constructor that uses `HashSet`s only to be serialized. Please see:

https://github.com/jenkinsci/folder-auth-plugin/blob/047102ace11ba532c2eb360be4543793a4fb7d17/src/main/java/io/jenkins/plugins/folderauth/FolderBasedAuthorizationStrategy.java#L66-L75

Like @runzexia says, some users may invoke the REST methods in parallel which may lead to concurrent modification exceptions being thrown. Using a `ConcurrentHashMap$KeySetView` in the memory allows us to maintain a thread-safe set in memory but one that doesn't get persisted. Could you please let me know what you think of this?

Reply:
Again without reference to the particulars of this case: _usually_ this sort of issue is solved simply by synchronizing access to the whole object. That is tricky in the case of an object held as part of another top-level `Saveable`, since you do not get a particular callback during serialization, so you would only be able to add synchronization around the mutations and accessors, not around the `save` lower in the call stack. Cannot recall offhand if there is a standard solution to that, since I cannot think of any cases where this has been a real problem.

Sorry to say that I cannot spend more time on this.